File: frontend/src/lib/api/schools.ts
import client from './client';

export interface SchoolBranding {
  school_name: string;
  short_name: string | null;
  school_type: 'BASIC' | 'SHS' | 'TECHNICAL' | 'VOCATIONAL' | 'PRIVATE';
  motto: string | null;
  logo_url: string | null;
  brand_color: string;
  school_code: string;
}

export interface SchoolRead {
  id: string;
  name: string;
  short_name: string | null;
  school_code: string;
  school_type: 'BASIC' | 'SHS' | 'TECHNICAL' | 'VOCATIONAL' | 'PRIVATE';
  ownership: 'PUBLIC' | 'PRIVATE';
  region_id: string;
  district_id: string;
  phone: string | null;
  email: string | null;
  address: string | null;
  motto: string | null;
  logo_path: string | null;
  logo_url: string | null;
  established_year: number | null;
  is_active: boolean;
  has_boarding: boolean;
  has_period_attendance: boolean;
  subdomain: string | null;
  custom_domain: string | null;
  brand_color: string;
}

/** SchoolRead plus usage stats — the superadmin dashboard list only. */
export interface SchoolSummary extends SchoolRead {
  student_count: number;
  staff_count: number;
  last_login_at: string | null;
}

/** PATCH /schools/me — self-service. Deliberately no subdomain/custom_domain
 * field: the backend rejects them here with 403 regardless (services/
 * school.py::update_school's allow_domain_change gate) — only a platform
 * superadmin can change a school's sign-in link, via the separate
 * superadmin school-edit form. */
export interface SchoolUpdatePayload {
  name?: string;
  short_name?: string;
  phone?: string;
  email?: string;
  address?: string;
  motto?: string;
  established_year?: number;
  has_boarding?: boolean;
  has_period_attendance?: boolean;
  brand_color?: string;
}

/** Extends SchoolBranding with routing identifiers returned for custom domain resolution. */
export interface SchoolByDomainResult extends SchoolBranding {
  school_code: string;
  subdomain: string | null;
}

export interface RegionRead {
  id: string;
  name: string;
  code: string;
}

export interface DistrictRead {
  id: string;
  name: string;
  code: string;
  region_id: string;
}

/** POST /schools payload — superadmin only. subdomain left blank auto-generates
 * one from `name` (services/school.py::create_school). */
export interface SchoolCreatePayload {
  name: string;
  short_name?: string;
  school_code: string;
  school_type: 'BASIC' | 'SHS';
  region_id: string;
  district_id: string;
  phone?: string;
  email?: string;
  address?: string;
  motto?: string;
  established_year?: number;
  subdomain?: string;
  has_boarding?: boolean;
}

/** PATCH /schools/{id} payload — superadmin only. school_code/school_type
 * are immutable after creation (not accepted by the backend at all). */
export interface SchoolAdminUpdatePayload {
  name?: string;
  short_name?: string;
  phone?: string;
  email?: string;
  address?: string;
  motto?: string;
  established_year?: number;
  has_boarding?: boolean;
  subdomain?: string;
  custom_domain?: string;
  is_active?: boolean;
}

/** Public endpoint — no auth required. */
export async function getSchoolBranding(subdomain: string): Promise<SchoolBranding> {
  const { data } = await client.get<SchoolBranding>(`/schools/public/${subdomain}`);
  return data;
}

/**
 * Resolve a custom domain to its school branding + routing identifiers.
 * Public endpoint — no auth required. Called before login on custom-domain deployments.
 */
export async function getSchoolByDomain(hostname: string): Promise<SchoolByDomainResult> {
  const { data } = await client.get<SchoolByDomainResult>('/schools/by-domain', {
    params: { h: hostname },
  });
  return data;
}

/** Authenticated endpoint — returns the current user's school branding. */
export async function getMySchoolBranding(): Promise<SchoolBranding> {
  const { data } = await client.get<SchoolBranding>('/schools/my-branding');
  return data;
}

/** Return full school profile for the authenticated user's school. */
export async function getMySchool(): Promise<SchoolRead> {
  const { data } = await client.get<SchoolRead>('/schools/me');
  return data;
}

/** Partially update the authenticated user's school profile. */
export async function updateMySchool(payload: SchoolUpdatePayload): Promise<SchoolRead> {
  const { data } = await client.patch<SchoolRead>('/schools/me', payload);
  return data;
}

/** Upload or replace the school logo. */
export async function uploadMyLogo(file: File): Promise<SchoolRead> {
  const form = new FormData();
  form.append('file', file);
  const { data } = await client.post<SchoolRead>('/schools/me/logo', form, {
    headers: { 'Content-Type': 'multipart/form-data' },
  });
  return data;
}

// ── Superadmin — school onboarding ──────────────────────────────────────────

/** List every school on the platform. Superadmin only. */
export async function listSchools(
  params?: { active_only?: boolean; search?: string },
): Promise<SchoolSummary[]> {
  const { data } = await client.get<SchoolSummary[]>('/schools', {
    params: { limit: 500, ...params },
  });
  return data;
}

/** Permanently delete a school. Superadmin only — the backend only ever
 * allows this for an already-disabled, genuinely empty school (zero
 * students, zero staff); see services/school.py::delete_school. */
export async function deleteSchool(id: string): Promise<void> {
  await client.delete(`/schools/${id}`);
}

/** Register a new school. Superadmin only — Tagnatek provisions each school. */
export async function createSchool(payload: SchoolCreatePayload): Promise<SchoolRead> {
  const { data } = await client.post<SchoolRead>('/schools', payload);
  return data;
}

/** Update any school's profile — including is_active, which is what
 * actually blocks sign-in (services/auth_lookup.py::resolve_school_id
 * excludes deactivated schools from every login/reset lookup). Superadmin only. */
export async function updateSchool(id: string, payload: SchoolAdminUpdatePayload): Promise<SchoolRead> {
  const { data } = await client.patch<SchoolRead>(`/schools/${id}`, payload);
  return data;
}

/** Public reference data — no auth required (also usable pre-login). */
export async function listRegions(): Promise<RegionRead[]> {
  const { data } = await client.get<RegionRead[]>('/schools/regions');
  return data;
}

export async function listDistricts(regionId?: string): Promise<DistrictRead[]> {
  const { data } = await client.get<DistrictRead[]>('/schools/districts', {
    params: regionId ? { region_id: regionId } : undefined,
  });
  return data;
}

const DEFAULT_FAVICON = '/favicon.svg';

/**
 * Apply the school's brand color as CSS custom properties on <html>, and
 * swap the browser-tab favicon to the school's own logo — the one piece of
 * "this feels like ours" branding that was still a static platform-wide
 * file (app.html's #app-favicon) even though title/logo/motto/color were
 * already dynamic everywhere this function is called (login, portal,
 * forgot-password, the live Setup preview). Falls back to the default
 * favicon when a school has no logo, rather than leaving a stale one from
 * whatever branding was applied previously (matters for ProfileTab's live
 * preview and for navigating between schools on the login screen).
 */
export function applyBranding(branding: SchoolBranding) {
  const root = document.documentElement;
  const hex = branding.brand_color;
  root.style.setProperty('--brand', hex);
  root.style.setProperty('--brand-light', hexTint(hex, 0.9));

  const favicon = document.getElementById('app-favicon') as HTMLLinkElement | null;
  if (favicon) favicon.href = branding.logo_url ?? DEFAULT_FAVICON;

  // RGB triplet for rgba(var(--brand-rgb), 0.1) usage in components
  const r = parseInt(hex.slice(1, 3), 16);
  const g = parseInt(hex.slice(3, 5), 16);
  const b = parseInt(hex.slice(5, 7), 16);
  root.style.setProperty('--brand-rgb', `${r}, ${g}, ${b}`);
}

function hexTint(hex: string, lightness: number): string {
  const r = parseInt(hex.slice(1, 3), 16);
  const g = parseInt(hex.slice(3, 5), 16);
  const b = parseInt(hex.slice(5, 7), 16);
  const tr = Math.round(r + (255 - r) * lightness);
  const tg = Math.round(g + (255 - g) * lightness);
  const tb = Math.round(b + (255 - b) * lightness);
  return `rgb(${tr}, ${tg}, ${tb})`;
}

File: backend/app/routers/schools.py
"""
Schools router — HTTP endpoints for school management.

This file is intentionally thin.  Its only jobs are:
  1. Declare the HTTP routes.
  2. Enforce access control via dependencies.
  3. Delegate to services/school.py for business logic.

ACCESS CONTROL SUMMARY
----------------------
GET  /schools/regions              Public — no auth required (needed for signup forms)
GET  /schools/districts            Public — no auth required
GET  /schools/public/{subdomain}   Public — branding for the login screen
GET  /schools/my-branding          Authenticated — returns branding for caller's school
GET  /schools/me                   Authenticated — full school profile for caller's school
GET  /schools/me/positions         Authenticated — list positions for caller's school
PATCH /schools/me                  Requires 'school.edit' permission (subdomain/custom_domain rejected — superadmin only, see PATCH /schools/{id})
POST /schools/me/logo              Requires 'school.edit' permission
POST /schools                      Superadmin only (require_superadmin)
GET  /schools                      Superadmin only — lists every school platform-wide
GET  /schools/{id}                 Superadmin only (legacy — use /me for self-service)
PATCH /schools/{id}                Superadmin only (legacy — use /me for self-service)
DELETE /schools/{id}               Superadmin only — narrow cleanup tool, see services/school.py::delete_school
POST /schools/{id}/logo            Superadmin only
PUT  /schools/{id}/config          Superadmin only (legacy — use /me self-service where it exists)
GET  /schools/{id}/config          Superadmin only
PUT  /schools/{id}/sms-config      Superadmin only

NOTE: the {id}-path endpoints above used to be gated by require_auth /
require_permission("school", "edit") — that only checks the CALLER's own
permission/school, never that the path {id} matches the caller's own
school_id, so any staff member holding school.edit at their own school
could read/edit ANY other school's profile, config, or SMS credentials by
substituting a different UUID in the path. Since the frontend only ever
calls the /me equivalents (confirmed: no {id}-path call site exists), these
are now superadmin-only — the same fix already applied to POST {id}/logo.
"""
from __future__ import annotations

import uuid

from fastapi import APIRouter, Depends, Query, UploadFile
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.database import get_db
from app.core.dependencies import require_auth, require_permission, require_superadmin
from app.models.auth import StaffPosition
from app.models.school import GhanaDistrict, GhanaRegion
from app.schemas.staff import PositionRead
from app.schemas.school import (
    DistrictRead,
    RegionRead,
    SchoolBranding,
    SchoolByDomainResult,
    SchoolConfigSet,
    SchoolCreate,
    SchoolRead,
    SchoolSummary,
    SchoolUpdate,
    SmsConfigCreate,
    SmsConfigRead,
)
from app.services import school as school_svc
from app.services import school_config as config_svc

router = APIRouter(prefix="/schools", tags=["schools"])


# ── Public reference data ─────────────────────────────────────────────────────
# These endpoints are intentionally unauthenticated because they are needed
# on school registration forms before any user account exists.

@router.get("/regions", response_model=list[RegionRead])
async def list_regions(db: AsyncSession = Depends(get_db)):
    """Return all 16 Ghana regions ordered by name."""
    rows = await db.scalars(select(GhanaRegion).order_by(GhanaRegion.name))
    return [RegionRead.model_validate(r) for r in rows]


@router.get("/districts", response_model=list[DistrictRead])
async def list_districts(
    region_id: uuid.UUID | None = Query(None, description="Filter by region"),
    db: AsyncSession = Depends(get_db),
):
    """Return Ghana districts, optionally filtered to a single region."""
    stmt = select(GhanaDistrict).order_by(GhanaDistrict.name)
    if region_id:
        stmt = stmt.where(GhanaDistrict.region_id == region_id)
    rows = await db.scalars(stmt)
    return [DistrictRead.model_validate(d) for d in rows]


# ── Public branding (no auth — called before login) ──────────────────────────

@router.get("/public/{subdomain}", response_model=SchoolBranding)
async def get_school_branding(
    subdomain: str,
    db: AsyncSession = Depends(get_db),
):
    """
    Return branding data for a school identified by its subdomain.

    No authentication required — this is called by the frontend before
    the login screen renders so the school's logo, colours, and motto
    can be displayed to the user before they log in.
    """
    return await school_svc.get_school_branding(subdomain, db)


# ── Custom domain resolution (no auth — called before login) ─────────────────

@router.get("/by-domain", response_model=SchoolByDomainResult)
async def get_school_by_custom_domain(
    h: str = Query(..., description="Full hostname, e.g. portal.presec.com"),
    db: AsyncSession = Depends(get_db),
):
    """
    Resolve a custom domain to its school and return branding + routing identifiers.

    No authentication required — called on mount when the browser is on a
    school-owned domain (e.g. portal.presec.com) rather than the platform
    subdomain (presec.ttek-sms.com).
    """
    return await school_svc.get_school_by_custom_domain(h, db)


# ── Authenticated branding (after login, school not yet in localStorage) ───────

@router.get("/my-branding", response_model=SchoolBranding)
async def get_my_school_branding(
    ids: tuple = Depends(require_auth),
    db: AsyncSession = Depends(get_db),
):
    """Return branding data for the currently authenticated user's school."""
    _user_id, school_id = ids
    return await school_svc.get_school_branding_by_id(school_id, db)


@router.get("/me", response_model=SchoolRead)
async def get_my_school(
    ids: tuple = Depends(require_auth),
    db: AsyncSession = Depends(get_db),
):
    """Return full school profile for the authenticated user's school."""
    _user_id, school_id = ids
    return await school_svc.get_school(school_id, db)


@router.get("/me/positions", response_model=list[PositionRead])
async def list_my_positions(
    ids: tuple = Depends(require_auth),
    db: AsyncSession = Depends(get_db),
):
    """Return all staff positions available to the authenticated user's school.

    School-specific positions take precedence over platform templates with the same code.
    """
    from sqlalchemy import or_
    _, school_id = ids
    rows = await db.scalars(
        select(StaffPosition)
        .where(or_(StaffPosition.school_id == school_id, StaffPosition.school_id.is_(None)))
        .order_by(StaffPosition.name)
    )
    # TEACHER, CLASS_TEACHER, and HOUSEMASTER are all derived (see
    # core/permissions.py::resolve_permissions), not manually granted —
    # exclude them from the manual "Authority" picker. TEACHER specifically
    # is the core role, not an optional responsibility like the other two.
    DERIVED_CODES = {"TEACHER", "CLASS_TEACHER", "HOUSEMASTER"}
    seen: dict[str, StaffPosition] = {}
    for pos in rows:
        if pos.code in DERIVED_CODES:
            continue
        if pos.code not in seen or pos.school_id is not None:
            seen[pos.code] = pos
    return sorted(seen.values(), key=lambda p: p.name)


@router.patch("/me", response_model=SchoolRead)
async def update_my_school(
    req: SchoolUpdate,
    ids: tuple = Depends(require_permission("school", "edit")),
    db: AsyncSession = Depends(get_db),
):
    """Update the authenticated user's school profile fields. subdomain/
    custom_domain are rejected here (403) even if included in the request —
    only the superadmin route can change a school's sign-in link."""
    _user_id, school_id = ids
    return await school_svc.update_school(school_id, req, db)


@router.post("/me/logo", response_model=SchoolRead)
async def upload_my_school_logo(
    file: UploadFile,
    ids: tuple = Depends(require_permission("school", "edit")),
    db: AsyncSession = Depends(get_db),
):
    """
    Upload or replace the school logo.

    Accepts JPEG, PNG, or WebP up to 2 MB.  Stored as WebP, max 400 × 400 px.
    School admins with school.edit permission can manage their own logo.
    """
    _user_id, school_id = ids
    return await school_svc.upload_school_logo(school_id, file, db)


# ── School registration (superadmin only) ─────────────────────────────────────

@router.post("", response_model=SchoolRead, status_code=201)
async def create_school(
    req: SchoolCreate,
    _ids: tuple = Depends(require_superadmin),
    db: AsyncSession = Depends(get_db),
):
    """
    Register a new school on the platform.

    Only the platform superadmin can do this.  School headmasters cannot
    self-register — Tagnatek provisions each school.
    """
    return await school_svc.create_school(req, db)


# ── School queries (any authenticated user) ───────────────────────────────────

@router.get("", response_model=list[SchoolSummary])
async def list_schools(
    active_only: bool = Query(False, description="Exclude deactivated schools"),
    limit: int = Query(100, ge=1, le=500, description="Maximum results to return"),
    offset: int = Query(0, ge=0, description="Number of results to skip"),
    search: str | None = Query(None, description="Partial match against name or school_code"),
    _ids: tuple = Depends(require_superadmin),
    db: AsyncSession = Depends(get_db),
):
    """List every school on the platform, paginated, with usage stats.
    Superadmin only — a school's own users already have everything they
    need via /schools/me. Defaults to showing disabled schools too (unlike
    every other list endpoint's active_only default) since this is the
    superadmin's own management view — they need to find a disabled school
    to re-enable it, not just the ones currently in use."""
    return await school_svc.list_schools(
        db, active_only=active_only, limit=limit, offset=offset, search=search
    )


@router.get("/{school_id}", response_model=SchoolRead)
async def get_school(
    school_id: uuid.UUID,
    _ids: tuple = Depends(require_superadmin),
    db: AsyncSession = Depends(get_db),
):
    """Fetch a single school by its UUID. Superadmin only — use /schools/me
    for the caller's own school."""
    return await school_svc.get_school(school_id, db)


# ── School updates (superadmin only — see module docstring) ──────────────────

@router.patch("/{school_id}", response_model=SchoolRead)
async def update_school(
    school_id: uuid.UUID,
    req: SchoolUpdate,
    _ids: tuple = Depends(require_superadmin),
    db: AsyncSession = Depends(get_db),
):
    """Update school profile fields.  school_code and school_type cannot be changed.
    Only this superadmin-only route may change subdomain/custom_domain — see
    services/school.py::update_school's own docstring for why."""
    return await school_svc.update_school(school_id, req, db, allow_domain_change=True)


@router.delete("/{school_id}", status_code=204)
async def delete_school(
    school_id: uuid.UUID,
    _ids: tuple = Depends(require_superadmin),
    db: AsyncSession = Depends(get_db),
):
    """Permanently delete a school — only ever safe for an already-disabled,
    genuinely empty one (zero students, zero staff). See services/school.py
    ::delete_school's own docstring for the full precondition list; there is
    no override for either check, from any caller."""
    await school_svc.delete_school(school_id, db)


@router.post("/{school_id}/logo", response_model=SchoolRead)
async def upload_logo(
    school_id: uuid.UUID,
    file: UploadFile,
    _ids: tuple = Depends(require_superadmin),
    db: AsyncSession = Depends(get_db),
):
    """
    Upload or replace the school logo.

    Superadmin only — logo is set during school provisioning, not by school admins.
    Accepts JPEG, PNG, or WebP up to 2 MB.  Stored as WebP, max 400 × 400 px.
    """
    return await school_svc.upload_school_logo(school_id, file, db)


@router.put("/{school_id}/config", status_code=204)
async def set_config(
    school_id: uuid.UUID,
    req: SchoolConfigSet,
    _ids: tuple = Depends(require_superadmin),
    db: AsyncSession = Depends(get_db),
):
    """Create or update a single config key-value pair for the school."""
    await config_svc.set_config(school_id, req, db)


@router.get("/{school_id}/config")
async def get_config(
    school_id: uuid.UUID,
    _ids: tuple = Depends(require_superadmin),
    db: AsyncSession = Depends(get_db),
):
    """Return all config key-value pairs for the school as a JSON object."""
    return await config_svc.get_config(school_id, db)


@router.put("/{school_id}/sms-config", response_model=SmsConfigRead)
async def upsert_sms_config(
    school_id: uuid.UUID,
    req: SmsConfigCreate,
    _ids: tuple = Depends(require_superadmin),
    db: AsyncSession = Depends(get_db),
):
    """Create or update the SMS provider configuration for the school."""
    return await config_svc.upsert_sms_config(school_id, req, db)
